Add UI for editing place types

tiltec opened the pull request:
Branch deployment URL: https://add-place-type-edit.dev.karrot.world

A nice UI for editing place types, very similar to the one for activity types.

Place type can be selected in place edit form. Custom icons will be shown in the place list and on the group map.

This lays the groundwork for refining display of places.

## Edit place types
![image](https://user-images.githubusercontent.com/4410802/189503543-01e6e5f6-9068-4aa0-bed1-4f434c5b8562.png)

![image](https://user-images.githubusercontent.com/4410802/189503558-bb626693-743f-4065-aade-4824d019c5d7.png)

## Select type for place
![image](https://user-images.githubusercontent.com/4410802/189515947-6e8ebd02-67f4-4499-81e6-5a475563242d.png)

## place type markers on map

![image](https://user-images.githubusercontent.com/4410802/189516094-02d06bc2-b101-4012-b55f-a107bdaa8fc3.png)

## notes

- custom icons might clash with pre-defined icons for users and groups
- place statuses are untouched, they still read "cooperating" etc...

Based on backend PR https://github.com/karrot-dev/karrot-backend/pull/1159

Related to https://github.com/karrot-dev/karrot-frontend/issues/2244

nicksellen commented:
Looks great!

> favorite places are not at top anymore

I think this is problematic, as groups with a lot of places likely use this a lot to keep it manageable.

However, I think with the new "explore places" UI it probably resolves this... I imagine for that one the favourites would stay on the sidebar?

> might be nice to let users specify the place type order in sidenav

I think alphabetical ordering is OK here, as groups already use that convention, currently they are not ordered that way.

**Suggestion: always order alphabetically by name** (translated name)

> it's a bit weird with singular names as headers in sidenav

It didn't strike me as odd before I read your comment, and even after, seems OK to my eye/brain...

> custom icons might clash with pre-defined icons for users and groups

yes, so be I think.... users can handle that. if you thought it more a problem, we could filter those out, as the icon picker let's us choose exactly which ones to show...

> sidenav feels a bit full now, maybe it's time for the map to go away?

I suspect the place UI re-jig will resolve that? although I agree in general, sidenav gets more and more stuff in it. but I think solving it is maybe out of scope of this, I would be in favour of going to more customizable level (on group and user levels).

> place statuses are untouched, they still read "cooperating" etc...

yes, indeed. it's a bit odd for non-foodsaving groups at the moment to see "co-operating"... I suspect we could just relabel that one to "active" (as I think the backend value is?), and it would be ok? ... and the custom place statuses is on the cards too... 

either way, out of scope of this one, but nice to remember it...

In addition I thought it would be nice to use the green colour when editing place types, as the green is a kind of "normal" state...

![Screenshot_2022-09-12_12-42-33](https://user-images.githubusercontent.com/31616/189646568-1f8b5c06-06aa-4e2f-88fa-a712683946e9.png)

I noticed it starts in "default" place type I think, with a circle icon, if I then select a place type, I can't get back to "default" or "no place type".... but I can't save without one set anyway... although the validation doesn't validate that (get server error). Minimal change is probably just adding it to the validation.

![Screenshot_2022-09-12_12-43-13](https://user-images.githubusercontent.com/31616/189646640-69bcb0cb-2828-49da-a6de-fcb697d1c97f.png)

I also wonder if we should _officially_ acknowledge the use of places for non-geographically purposes, and include a translatable name for: "Discussion" or "Working Group"? (although requires the backend change to also have those names I think).

Can throw some of these questions into the weekly meeting tomorrow (with or without attendance) if you'd like any decisions including the others :)

nicksellen commented:
Oh, and the icon for "place type" in the form is an eye, presumably a copy and paste error? I actually started removing some of these icons on forms, I'm not convinced they help, and clutter up the form and layout on small screens/mobile.

nicksellen commented on src/group/pages/EditPlaceTypes.spec.js:
```diff
@@ -0,0 +1,89 @@
+    await click(await findByTitle('Add new place type'))
+    const textbox = await findByRole('combobox', { name: 'Name' })
+    // somehow we need to type the name before we can choose it
+    // seems like a problem between jsdom/testing-library and Quasar
```
My guess is that because the select is closed at first, you need to click the `findByRole('combobox', { name: 'Name' })` to open it... then can click the Distribution Place within it?

tiltec replied:
the comment is also copy-pasta and likely wrong. in this case we really need to type the name as it's not a predefined option.

nicksellen commented on src/locales/locale-en.json:
```diff
@@ -684,6 +685,32 @@
+    "CUSTOM_NAME_USE": "Use custom name {name}",
+    "CUSTOM_NAME_PROMPT": "Type a custom name, or choose a standard name",
+    "MANAGE_TYPES": "Manage types...",
+    "SHOW_ARCHIVED": "Show archived",
```
Many of these translations are duplicate from activity types, I have been tending to re-use things in this case, even if the i18n keypath is a bit nonsensical... and then imagining if we can migrate keypaths properly one day (did we manage this before?)

tiltec replied:
They are duplicated, yeah.. I think it's not too bad, these small messages are fast to translate. Transifex has some hints when a string was already translated, you can just click and it fills the previous one.

nicksellen commented on src/places/components/PlaceEdit.vue:
```diff
@@ -25,6 +25,55 @@
+            class="q-mb-lg"
+          >
+            <template #before>
+              <QIcon name="fas fa-eye" />
```
This is the copy-and-paste eye I guess? ... and my suggestion is actually to remove all the #before icons....

nicksellen commented on src/places/components/PlaceEdit.vue:
```diff
@@ -265,8 +318,26 @@ export default {
+
+    const placeTypes = computed(() => getPlaceTypesByGroup(groupId))
+
+    // TODO sorting?
```
Yes please!

nicksellen commented on src/places/mutations.js:
```diff
@@ -38,3 +40,28 @@ export function useSavePlaceMutation () {
   ))
 }
+
+// TODO what about websocket place type updates?
```
Could be a nice backend addition, although not expecting them to change a lot, so maybe OK. And could make the query re-load them here and there with the right options...

nicksellen commented on src/places/queries.js:
```diff
@@ -51,3 +53,19 @@ export function usePlaceListQuery (queryOptions = {}) {
+    () => placeTypeAPI.list(),
+    {
+      placeholderData: () => [],
+      staleTime: Infinity,
```
Could rjiggle this to cause them to be reloaded now and again, given there are no socket updates... (if I understood correctly). Not sure the best options for that.

nicksellen reviewed (approved):
Nice, great work! I added some suggestions for a few changes here and there, but approved in general.

tiltec commented:
I removed the place types in the sidenav, as this is going to change in the refine-place-display PR. Which resolves approximately half of the comments ;)

I also removed the icons in PlaceEdit. looks a bit leaner now.

I just noticed a small backend quirk: if all place types are disabled and I create a new place without type, the backend picks the "Unspecified" type, presumably as default.

> I noticed it starts in "default" place type I think, with a circle icon, if I then select a place type, I can't get back to "default" or "no place type".... but I can't save without one set anyway... although the validation doesn't validate that (get server error). Minimal change is probably just adding it to the validation.

@nicksellen not sure I follow - I can save successfully without place type (backend auto-fills the default), but maybe I shouldn't be able to? Can add validation, sure :)

> When creating a new group empty of places, nothing happens when clicking the plus button to create a place

fixed (by undoing the sidenav)

> I agree with Nick in adding "Discussions" or "Working group". Not only because it's an already existing use case, but also because it's a nice way to point people towards this possibility.

@brnsolikyl these pre-defined values also need to be added in the backend, so let's do this another time?
